src/main/java/com/internet/shop/dao/mysqlimpl/UserDaoJdbcImpl.java

```diff
@@ -0,0 +1,147 @@
+package com.internet.shop.dao.mysqlimpl;
+
+import com.internet.shop.dao.interfaces.UserDao;
+import com.internet.shop.exceptions.DataProcessingException;
+import com.internet.shop.lib.Dao;
+import com.internet.shop.model.Role;
+import com.internet.shop.model.User;
+import com.internet.shop.util.ConnectionUtil;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+@Dao
+public class UserDaoJdbcImpl implements UserDao {
+
+    @Override
+    public Optional<User> findByLogin(String login) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "SELECT * FROM users WHERE login=?;";
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setString(1, login);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                return Optional.of(getUserFromResultSet(resultSet, connection));
+            }
+            return Optional.empty();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get user by login " + login, e);
+        }
+    }
+
+    @Override
+    public User create(User user) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "INSERT INTO users (user_name, login, password) VALUES (?,?,?)";
+            PreparedStatement statement = connection
+                    .prepareStatement(query, Statement.RETURN_GENERATED_KEYS);
+            statement.setString(1, user.getName());
+            statement.setString(2, user.getLogin());
+            statement.setString(3, user.getPassword());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                user.setId(resultSet.getLong(1));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can not create user " + user, e);
+        }
+        return user;
+    }
+
+    @Override
+    public Optional<User> get(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "SELECT * FROM users WHERE user_id=?;";
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                return Optional.of(getUserFromResultSet(resultSet, connection));
```

Comment:
same with two opened statements, check all methods in your PR for this problem

Reply:
i do not understand where second statement opens actually

Reply:
First statement opens on 64 line and the next one on 155 line inside getUserRoles(). You may close the first connection here using method close().  Look for the all similar situation  in you dao classes and fix it